src/shogun/statistical_testing/internals/mmd/ComputeMMD.h

```diff
@@ -108,13 +108,14 @@ struct ComputeMMD
 		const BlockXt& b_xy=map.block(m_n_x, 0, m_n_y, m_n_x);
 
 		terms_t terms;
-		terms.diag[0]=b_x.diagonal().sum();
-		terms.diag[1]=b_y.diagonal().sum();
-		terms.diag[2]=b_xy.diagonal().sum();
+		// cast to 64 bit to avoid overflows
+		terms.diag[0]=b_x.diagonal(). template cast<float64_t>().sum();
+		terms.diag[1]=b_y.diagonal(). template cast<float64_t>().sum();
+		terms.diag[2]=b_xy.diagonal(). template cast<float64_t>().sum();
 
-		terms.term[0]=(b_x.sum()-terms.diag[0])/2+terms.diag[0];
-		terms.term[1]=(b_y.sum()-terms.diag[1])/2+terms.diag[1];
-		terms.term[2]=b_xy.sum();
+		terms.term[0]=(b_x. template cast<float64_t>().sum()-terms.diag[0])/2+terms.diag[0];
+		terms.term[1]=(b_y. template cast<float64_t>().sum()-terms.diag[1])/2+terms.diag[1];
+		terms.term[2]=b_xy. template cast<float64_t>().sum();
```

Comment:
so is this happening without copying the matrix?

Reply:
idk :D

Reply:
since I don't store the result in a variable, `eval` should not be called, and the `sum` should cast element by element in the loop. You have an idea how to check this easily/quickly?

Reply:
I am not sure, from the docs it seems like you can force lazy evaluation by using `.noalias` when you store the temporary values (see https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html)

Reply:
Ah yes I remember reading that page. I understand it like this: These docs also say that everything is lazy apart from 3 cases which are all different to this one. So merging.